src/compiler/checker.ts

```diff
@@ -45578,6 +45637,297 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            if (initialType === typeParam) {
+                return undefined;
+            }
+            const flowType = getFlowTypeOfReference(narrowReference, initialType);
+            const exprType = getTypeFromFlowType(flowType);
+            // Don't narrow the return type if narrowing didn't produce a narrower type for the expression.
+            if (isTypeAny(exprType) || isErrorType(exprType) || exprType === typeParam || exprType === mapType(typeParam, getBaseConstraintOrType)) {
+                return undefined;
+            }
+            const narrowedType = getSubstitutionType(typeParam, exprType, /*isNarrowed*/ true);
+            return [typeParam, narrowedType];
+        });
+
+        const narrowMapper = createTypeMapper(narrowed.map(([tp, _]) => tp), narrowed.map(([_, t]) => t));
+        const narrowedReturnType = instantiateType(
+            unwrappedReturnType,
+            narrowMapper,
+        );
+
+        if (expr) {
+            const links = getNodeLinks(expr);
+            if (!links.contextualReturnType) {
+                links.contextualReturnType = narrowedReturnType;
+            }
+        }
+
+        const narrowedExprType = expr ? checkExpression(expr) : undefinedType;
+        const narrowedUnwrappedExprType = functionFlags & FunctionFlags.Async
+            ? checkAwaitedType(
+                narrowedExprType,
+                /*withAlias*/ false,
+                node,
+                Diagnostics.The_return_type_of_an_async_function_must_either_be_a_valid_promise_or_must_not_contain_a_callable_then_member,
+            )
+            : narrowedExprType;
+        checkTypeAssignableToAndOptionallyElaborate(narrowedUnwrappedExprType, narrowedReturnType, errorNode, effectiveExpr);
+    }
+
+    function checkConditionalReturnExpression(
+        container: SignatureDeclaration,
+        returnType: Type,
+        node: ReturnStatement | Expression,
+        expr: ConditionalExpression,
+    ): void {
+        checkReturnExpression(container, returnType, node, expr.whenTrue, checkExpression(expr.whenTrue), /*inConditionalExpression*/ true);
+        checkReturnExpression(container, returnType, node, expr.whenFalse, checkExpression(expr.whenFalse), /*inConditionalExpression*/ true);
+    }
+
+    // Narrowable type parameters are type parameters that:
+    // (1) have a union type constraint;
+    // (2) are used as the type of a single parameter in the function, and nothing else
+    function getNarrowableTypeParameters(candidates: TypeParameter[]): [TypeParameter, Symbol, Identifier][] {
+        const narrowableParams: [TypeParameter, Symbol, Identifier][] = [];
+        for (const typeParam of candidates) {
+            const constraint = getConstraintOfTypeParameter(typeParam);
+            if (!constraint || !(constraint.flags & TypeFlags.Union)) continue;
+            if (typeParam.symbol && typeParam.symbol.declarations && typeParam.symbol.declarations.length === 1) {

```

Comment:
It can be > 1. If you have multiple declarations of an interface, e.g. `Array<T>` in the lib files, then a type parameter can have multiple declarations. Not sure if there are other ways, but there's at least that one.